### DISCUSSION TITLE
Dual Editors - an Alternative 2.0 Architectural Proposal

### DISCUSSION THREAD

**Discussion by acao** [graphql/graphiql#1787] [Ideas]
The goals are similar to what we had last year, a robust ecosystem for developing web-based graphql development experiences with react and monaco. 

The main difference with this variant proposal is two editors instead of one, and though this may seem more impractical, I will attempt to prove why it may be even *more* practical. 

![image](https://user-images.githubusercontent.com/1368727/106987486-92e2ca80-673b-11eb-86a6-4934640ff02a.png)

Instead of working *directly* from the RFC, we could focus on building out `@graphiql/toolkit` and `@graphiql/react` by porting existing logic into new patterns and APIs. We can adapt components and libraries from graphiql 2 rfc or entirely rewrite many parts as needed.

`graphiql` would remain lightweight, potentially more than it is already, a reference implementation for docs & spec development. It would implement the same components and css-in-js theme customization system as the `graphql-ide`, and thus be included in the redesign of course.

Then, `graphql-ide` (tentative title) could move forward with introducing greater customizeability and extensibility without encumbering the tool we use for developing the reference ecosystem

- `@graphiql/toolkit` will become the new home of many of the most important `graphiql@1` business logic. everything from executing operations, to AST manipulations, etc. This logic lives within `src/utilities` or within giant components like `GraphiQL`
- `@graphiql/react` would export hooks for accessing state, theme system components, editor components, doc explorer?, query history, etc. some of these will be abstractions around `@graphiql/toolkit`. it will act as a reference implementation of `@graphiql/toolkit` that other frameworks like `svelte` or `vue`, etc could use for inspiration
- `@graphiql/create-fetcher` has been created to generate a fetcher, which can be used by both `graphiql` in the current pattern, and internally by `graphql-ide`
- `@graphiql/react-doc-explorer` - perhaps? though it could just be exported from the `@graphiql/react` package

thoughts?

**Comment by imolorhe** [comment]
I like the idea of having a `@graphiql/core` that contains most of the business logic and having the `@graphiql/react` abstraction as you have described.

I assume the create fetcher package is meant to be a utility to enable creating pluggable fetching logic for graphiql, however is the logic that complex that it needs to be a separate package? Another way of looking at the question: why does `create-fetcher` deserve a separate package and the other utilities in the core package do not?

A similar argument goes for `react-doc-explorer`. Why does it deserve a separate package and the other components (editor, result viewer, etc which can also be considered as having a good amount of complexity) do not?

About creating the separate editors, my take on it is: the architecture proposal is meant to cater for the maximum required level of flexibility and extensibility. If that goal is achieved, creating a simple build/component abstraction shouldn't be an issue any longer, as it will just be using the editor and specifying reasonably simple default options. So while technically you could say there are two editors, we would just have a single implementation with the second lean editor being a simple usage of the first.

**Reply by acao** [reply]
> A similar argument goes for react-doc-explorer. Why does it deserve a separate package and the other components (editor, result viewer, etc which can also be considered as having a good amount of complexity) do not?

yes this was all discussed at the end of the proposal, that really these don't need to be seperate packages

the fetcher probably should have been part of `@graphiql/toolkit` as well

but! luckily, we're still in RC stage for graphiql, so we can fix that

> , we would just have a single implementation with the second lean editor being a simple usage of the first.

 the way i see it, the new `graphiql` will just be the `@graphiql/react` getting started example, does that frame it differently?

**Reply by imolorhe** [reply]
>the way i see it, the new graphiql will just be the @graphiql/react getting started example, does that frame it differently

That's exactly how I was thinking of it in my head, but was careful but to call it just an "example" 😄 but that's the same thought I have as well

**Reply by acao** [reply]
for sure! i think i need to make this more clear in this proposal!

**Reply by imolorhe** [reply]
Just to clarify, the questions i asked about "why does component A deserve a separate package and the other components do not?", I wasn't saying it rhetorically but actually asking in the sense that, depending on the answer, it might still be a good idea to create separate packages as long as the benefits are properly outlined.

**Reply by acao** [reply]
yes thats how i interpreted it! I think here's a few reasons why i would argue *for* making create-fetcher a seperate package:
- manage a discreet set of related dependencies (tree shaking, but still)
- a single purpose makes it easier for transport spec folks to maintain
- easier to document

however, the toolkit is where we will be moving other `fetcher` related business logic anyways, so it may be most practical to bring it back into the toolkit

**Reply by acao** [reply]
I think for the next rc we should move it to toolkit

and, here’s something else thats going into toolkit!

https://github.com/graphql/graphiql/pull/1579

@harshithpabbati had rewritten it for this new architecture

i think i can make a roadmap just for toolkit, and we can make a list of what we want to pull in from where

**Reply by acao** [reply]
@imolorhe thinking we should actually move the fetcher to the `graphql-language-service` because it's already used by both `graphql-language-service-server` and `monaco-graphql`.

the reason being is that all of this would be fantastic to use for the language server, for `vscode-graphql`, etc as well

**Reply by acao** [reply]
we will move the fetcher to `@graphiql/toolkit` for now, and then if we want to move it to `graphql-language-service` later we can, and then we can export it from `@graphiql/toolkit` as an alias

**Reply by acao** [reply]
so, if some of the logic that's in `packages/graphiql` now, which will be moved to `@graphiql/toolkit` needs to be used in `vscode-graphql` extension client, I can just import `@graphiql/toolkit` ... that makes sense because they're both client side consumers.

it would go to `graphql-language-service` if it needs to work in node, browser main and webworker runtime

**Comment by airhorns** [comment]
This sounds super awesome -- I think the approach of a high level, batteries included component which you could "eject" from to build your own high level system out of the lower level bits makes a lot of sense!

**Comment by ryan-m-walker** [comment]
I really like that this approach also will allow people to potentially easily consume the many utilities and components for any other kind of graphql related projects they might have

**Reply by acao** [reply]
Yes that’s the goal! Hoping that instead of building complex APIs, we can instead build the editor with everything one would need to build their own editor.

The theme layer brings up some questions here though - what if someone wants to use a different theming system like chakra UI? They should be able to re-use some of the most powerful components that are potentially decoupled from the theming layer perhaps?

**Reply by imolorhe** [reply]
I'm not sure the theme should be configurable to the extent of choosing the theming system 🤔

**Reply by acao** [reply]
agreed, that's why there will be a configurable application using our chosen theming system, and a way to customize that theme using configuration, plugins/etc, but essentially one chosen theme system.

however, if one should want to "eject" from the configuration/plugin-driven `graphiql-ide` service layer, they might instead build your own application in react that uses the reusable `@graphiql/react` operation/variables editor components, some of the utilities, some of the hooks? and then implement their chosen theming system. you also might not be able to use plugins in the same way or at all if you "eject" to this extent.

the even more arduous and adventurous route would be to use `@graphiql/toolkit`, `@graphiql/create-fetcher`, and `monaco-graphql` to build an app in vue, svelte, angular, however! on that note, a `@graphiql/vue` or `@graphiql/svelte` would be quite welcome some day!

**Comment by acao** [comment]
Here is a PR that's a great starting point towards where we are headed with `@graphiql/react`, thanks to @ryan-m-walker :

https://github.com/graphql/graphiql/pull/1803

We also should move many of the graphiql utilities to `@graphql/toolkit` as part of this move

The end goal is that the rewritten `graphiql` 2, with modern react hooks, state management and theme system is a simplified end implementation of `@graphiql/react` and `@graphiql/toolkit`. referred to as "the sdk"

the playground-like project that will emerge from this effort, whatever it will be called, will be based on the same underlying components but will have more advanced features & settings, ways to provide simple "Plugin" components in the UI (custom forms, results panes, variables editors, query editors, etc). there will probably need to be two service layers essentially

the current pattern of just dumping props into a single react component will slow us down eventually. it won't enable the kinds of tools people want to build with graphiql or this ecosystem. it's become like the jquery mega-libraries of yore, with API, business and presentation concerns all jumbled together in one giant flat interface. the playground approach to settings via a simple json or yaml config makes way more sense, is more portable across service interfaces, and also might allow for some degree of parity in the transition for playground users, and for more advanced usages, direct use of the SDK and ample examples will serve them well

CC @dotansimha

**Comment by acao** [comment]
How do the original authors of Playground feel about this proposal - @timsuchanek @schickling @huv1k?

I see two potential paths forward, either this one where we re-build playground from the ground up ala 2.0 for monaco with this SDK approach 

-or-

We just set up a team to specifically focus on reviving stability & maintenance of the existing playground codebase, and port it to use `monaco-graphql` either by default or optionally depending on what users want for 2.0? 

There are users helping other users with issues & PRs over there who may make excellent co-maintainers!

We can enable discussions in the repository and let the users decide 🗳️
